Distributed materials

smharper opened the pull request:
This PR implements the distributed materials capability we discussed at the December meeting.  Here are some relevant inputs and plots from the test to show how it works:

```python
        c11.fill = [dense_fuel, light_fuel, 'void', dense_fuel]
```
```xml
    <cell id="11" material="2 3 void 2" region="-10000" universe="11" />
```
Cell plot:
![cellplot](https://cloud.githubusercontent.com/assets/4622797/12220059/25e8d722-b72c-11e5-9490-4072e671cc5b.png)

Material plot:
![matplot](https://cloud.githubusercontent.com/assets/4622797/12220061/306ef5e6-b72c-11e5-8d98-c18cdc665dfe.png)

Note that materials that go into an array will need to be input in the rather unintuitive order of z, y, x, but that's how distribcell offsets are calculated and I don't want to change that in this PR.

paulromano commented on docs/source/usersguide/output/summary.rst:
```diff
@@ -91,10 +91,12 @@ The current revision of the summary file format is 1.
+    Unique ID of the material(s) assigned to the cell. This dataset is present
+    only if fill_type is set to 'normal'.  The data is an array if the cell uses
+    distributed materials, otherwise it is a scalar.  The value '-1' signifies
+    void material.
```
I would qualify the last sentence by saying "When given as an array".

smharper replied:
The `-1`s are used for the scalar outputs as well.  I changed the wording order to make that more clear.

paulromano commented on openmc/universe.py:
```diff
@@ -50,7 +50,7 @@ class Cell(object):
     name : str
         Name of the cell
-    fill : Material or Universe or Lattice or 'void'
+    fill : Materials or Universe or Lattice or 'void'
```
Really should be "Material or Universe or Lattice or 'void' or Iterable of Material".

paulromano commented on src/initialize.F90:
```diff
@@ -953,58 +954,67 @@ subroutine prepare_distribcell()
 
+    ! Make sure the number of materials matches the number of cell instances for
+    ! distributed materials
+    if (count_all) then
```
You have three separate `if (count_all)` blocks preceded by a single `if (.not. count_all)`. Why not have a single if-else?

paulromano replied:
Sorry, I see that the `if (.not. count_all)` block can actually change `count_all` itself. However, the other three ifs can be combined, no?

paulromano commented on src/initialize.F90:
```diff
@@ -953,58 +954,67 @@ subroutine prepare_distribcell()
-            t%filters(j)%n_bins = c%instances
+          if (t % filters(j) % type == FILTER_DISTRIBCELL) then
+            ! Set the number of bins to the number of instances of the cell
+            c => cells(t % filters(j) % int_bins(1))
```
Can you get rid of the `t`, `c`, and `univ` pointers and just use associate constructs (or plain references to elements of the corresponding arrays)?

paulromano commented on src/initialize.F90:
```diff
@@ -953,58 +954,67 @@ subroutine prepare_distribcell()
-    if (count_all) then
-
-      univ => universes(BASE_UNIVERSE)
+    ! We also need to count instnaces if any distributed materials are present.
```
Typo: instnaces

paulromano commented on src/initialize.F90:
```diff
@@ -953,58 +954,67 @@ subroutine prepare_distribcell()
+            end if
+          end if
+        end associate
+      end do
```
Can you get rid of the explicit deallocate statements at the end of this subroutine? They are automatically deallocated per the standard.

paulromano commented on src/initialize.F90:
```diff
@@ -1062,6 +1073,14 @@ recursive subroutine allocate_offsets(univ_list, counts, found)
 
+    ! List all cells with multiple (distributed) materials.
+    do i = 1, n_cells
+      c => cells(i)
```
I would use associate construct instead.

paulromano commented:
Everything thing looks good save for my minor comments. Thanks for submitting a relatively small PR to get this started!

smharper commented:
I believe I've addressed all your comments.  I did a pretty big overhaul of `prepare_distribcell`.  It now uses `distribcell_active` instead of `count_all` and it returns early `if (.not. distribcell_active)`.  One side-effect of this is that `allocate_offsets` is no longer always called, but I don't think it needs to be.

paulromano commented on src/initialize.F90:
```diff
@@ -943,69 +945,71 @@ end subroutine allocate_banks
+      do j = 1, tallies(i) % n_filters
+        if (tallies(i) % filters(j) % type == FILTER_DISTRIBCELL) then
+          distribcell_active = .true.
+          if (size(tallies(i) % filters(j) %int_bins) > 1) then
```
Add space before `int_bins`

smharper replied:
Fixed

paulromano commented on openmc/summary.py:
```diff
@@ -477,10 +478,14 @@ def _finalize_geometry(self):
+                    if fill_id > 0:
+                        fill = self.get_material_by_id(fill_id)
+                    else:
+                        fill = 'void'
```
Might as well use ternary form `fill = self.get_material_by_id(fill_id) if fill_id > 0 else 'void'`

smharper replied:
But then I'd need a line continuation to keep it under 80 cols so I figure we may as well leave the verbose form.

paulromano replied:
Ah, so be it then

paulromano commented on src/initialize.F90:
```diff
@@ -614,31 +614,33 @@ subroutine adjust_indices()
         else
           call fatal_error("Specified fill " // trim(to_str(id)) // " on cell "&
-               &// trim(to_str(c%id)) // " is neither a universe nor a &
+               &// trim(to_str(c % id)) // " is neither a universe nor a &
```
Remove leading ampersand

paulromano commented on src/initialize.F90:
```diff
@@ -614,31 +614,33 @@ subroutine adjust_indices()
+            c % material(j) = material_dict % get_key(id)
+          else
+            call fatal_error("Could not find material " // trim(to_str(id)) &
+                 &// " specified on cell " // trim(to_str(c % id)))
```
Remove leading ampersand

paulromano commented on src/initialize.F90:
```diff
@@ -1036,38 +1034,33 @@ recursive subroutine allocate_offsets(univ_list, counts, found)
-    type(TallyObject), pointer :: t             ! pointer to tally
-    type(TallyFilter), pointer :: filter        ! pointer to filter
+    integer      :: i, j, k   ! Loop counters
+    type(SetInt) :: cell_list ! distribells to track
```
You should add a call to `cell_list % clear()` at the end of this subroutine. Otherwise we have a (small) memory leak.

paulromano commented on src/input_xml.F90:
```diff
@@ -1086,36 +1086,50 @@ subroutine read_geometry_xml()
+              ! Check for error
+              if (c % material(j) == ERROR_INT) then
+                call fatal_error("Invalid material specified on cell " &
+                     &// to_str(c % id))
```
Remove leading ampersand

paulromano commented on tests/test_distribmat/test_distribmat.py:
```diff
@@ -0,0 +1,128 @@
+#        plot.pixels = (400, 400)
+#        plots_file.add_plot(plot)
+#
+#        plots_file.export_to_xml()
```
I would either leave this code in un-commented or remove it altogether.

smharper replied:
I decided to un-comment it and add the plots.xml to the `_cleanup`.  People will likely look at this test as an example of how to use the distribmat capability, and I think leaving in the plots makes it easier to learn from.

paulromano commented:
Ok, I think that was my last round of comments for real.

smharper commented:
That last commit should address all your comments

paulromano commented:
Looks like you need to update inputs_true.dat for test_distribmat.

smharper commented:
I don't understand that failure.  plots.xml isn't included in the input hash.  And it passes on my computer

smharper commented:
Oh, I that changed in #547.  The latest commit should make it all better.
